### ISSUE TITLE
404

### ISSUE THREAD

**Issue by fabienjuif** [ptitFicus/naive-router#72]
Hey @ptitFicus 👋 ,

Do you have an idea to handle 404?
I think having a router and register routes to it is the only solution.

I had this in mind:
- Each route register itself to a **Router** using an effect to unregister itself at unmount
- Router register itself to window events (like you did there: https://github.com/ptitFicus/naive-router/blob/master/src/Route.js#L24)
- For each event, the router calls the route (via dispatch)

## Advantages
1. Only one listener per event
2. Handle 404
3. Move all logic in Router

## Drawbacks
1. A lot of reworking :(
2. Move all logic in Router

### Router
```js
const router = () => {
  routes = []
  
  const unregister = (path, callback) => () => {
    routes.remove(path, callback)
  }

  const register = (path, callback) =>{
    routes.set(path, callback)
    return unregister(path, callback)
  }

  const register404 = () => { /* */ }

  const dispatch = () => {
    let found = false
    for (route of routes) {
      if (/* route match new path */) {
        found = true
        route.callback(true)
      }
      route.callback(false)
    }

    if (!found) {
      /* calls 404 routes callbacks */
    }
  }

  window.on("push", dispatch)
  window.on("replace", dispatch)
  // ...

  return {
    register,
    register404,
  }
}
```

### Route
```js
const Route = ({ path, children }) => {
  const [matched, setMatched] = useState(false)
  const router = useContext(RouterContext)

  useLayoutEffect(() => {
    return router.register(path, setMatched)
  }, [])

  if (matched) return children
  return null
}
```

### Route404
```js
const Route404 = ({ children }) => {
  const [matched, setMatched] = useState(false)
  const router = useContext(RouterContext)

  useLayoutEffect(() => {
    return router.register404(path, setMatched)
  }, [])

  if (matched) return children
  return null
}
```

**Comment by ptitFicus** [comment]
Hi @fabienjuif, thank you for your proposal !

404 is instead an identified missing feature of this lib. I think it's indeed necessary in most project.

I think it may be possible without using a `Router` component, by making each route return wether or not it matched the path and implementing a logic in EventQueue  if no route has matched.

I'll try to implement it this week :)

**Comment by fabienjuif** [comment]
Thank you, this is even better indeed!

I am waiting your PR. I want to try the lib for a POC for one of our client 😊

**Comment by ptitFicus** [comment]
Ok I've done something, could you take a look at #73 ?

**Comment by ptitFicus** [comment]
Solved by #73
